fuseimg/datasets/duke.py

```diff
@@ -0,0 +1,389 @@
+             dict(key_in_sequence_ids='data.input.sequence_ids',
+                  key_in_path_prefix='data.input.path.',
+                  key_in_volumes_prefix='data.input.volumes.',
+                  key_out_paths='data.input.selected_paths',
+                  key_out_volumes='data.input.selected_volumes')),
+
+            # step 6: set reference volume to be first and register other volumes with respect to it
+            (ops_mri.OpResampleStkVolsBasedRef(reference_inx=0, interpolation='bspline'),
+             dict(key='data.input.selected_volumes')),
+
+            # step 7: create a single 4D volume from all the sequences (4th channel is the sequence)
+            (ops_mri.OpStackList4DStk(delete_input_volumes=True), dict(key_in='data.input.selected_volumes',
+                                                                           key_out_volume4d='data.input.volume4D',
+                                                                           key_out_ref_volume='data.input.ref_volume')),
+
+            # step 8:
+            (ops_mri.OpRescale4DStk(), dict(key='data.input.volume4D')),
+
+            # step 9: read tabular data for each patch
+            (ops_mri.OpAddPatchesData(get_annotations_func=get_annotations),
+             dict(key_out='data.input.patch_tab_data')),
+
+            # step 10: create patch volumes: (i) fixed size around center of annotatins (orig), and (ii) entire annotations
+            (ops_mri.OpCreatePatchVolumes(lsn_shape=(9, 100, 100), lsn_spacing=(1, 0.5, 0.5), delete_input_volumes=False),
+             dict(key_in_volume4D='data.input.volume4D',
+                  key_in_ref_volume='data.input.ref_volume',
+                  key_in_patch_row='data.input.patch_tab_data',
+                  key_out_cropped_vol='data.input.patch_volume_orig',
+                  key_out_cropped_vol_by_mask='data.input.patch_volume'))
+        ]
+        static_pipeline = PipelineDefault("static", static_pipeline_steps, verbose=verbose)
+
+        return static_pipeline
+
+    @staticmethod
+    def dynamic_pipeline(label_type: Optional[DukeLabelType]=None, verbose:Optional[bool]=True,
+                        include_patch_by_mask:Optional[bool]=True, include_patch_fixed:Optional[bool]=False):
+        assert include_patch_by_mask or include_patch_fixed
+        volume_keys = []
+        if include_patch_by_mask:
+            volume_keys.append('data.input.patch_volume')
+        if include_patch_fixed:
+            volume_keys.append('data.input.patch_volume_orig')
+
+        dynamic_steps = [ (ops_mri.OpDeleteLastChannel(), dict(keys=volume_keys)) ,
+                          (ops_mri.OpStk2Torch(), dict(keys=volume_keys))]
+        keys_2_keep = list(volume_keys)
+        if label_type is not None: #todo: wasn't checked!!!
+            key_ground_truth = 'data.ground_truth'
+            dynamic_steps.append((OpAddDukeLabelAndClinicalFeatures(label_type=label_type),
```

Comment:
It seems to me like OpConcat - op that getting a list of clinical features and concat them into one single vector.

Reply:
Need to revisit clinical features in Duke. I did not check with, and suggest to currently disable